Error handling improvements in chunk build threads

comp500 opened the pull request:
I noticed in some recent issue reports that chunk build exceptions weren't being propagated to the main thread as intended in https://github.com/CaffeineMC/sodium-fabric/pull/615. This is because deferred (non-important) chunk builds aren't polled from the main thread using `join()` unlike important chunk rebuilds, and since https://github.com/CaffeineMC/sodium-fabric/commit/696da3c219b0d21febdea9dd73e07516f694248a Sodium always treats initial chunk builds as deferred, so if a particular error makes a chunk never successfully build, `join()` will never be called on it!

This PR therefore correctly propagates these exceptions, along with extra handling for crash reporting that adds Sodium rendering context, with the following changes:

- Added a queue to store chunk build failures in `ChunkBuilder`
- Modified `scheduleDeferred` to add `CompletableFuture` exceptions to this queue
- Created `handlePendingFailures` in `RenderSectionManager` to drain the failure queue and throw any found exception
- Modified `WorkStealingFutureDrain.findNext` to unwrap `CompletionException`s that contain `CrashException` so that the crash report context is preserved
- Wrapped the `ChunkRenderRebuildTask.performBuild` outer loop in a try-catch block to handle exceptions from chunk building and add chunk section / render context information
- Improved `RenderSection.toString` to provide more useful information
- Changed `ChunkBuilder.processJob` to use the Sodium logger rather than `printStackTrace`

For `scheduleDeferred` I changed it to use `whenComplete` as it doesn't wrap the exception in `CompletionException` (so any `CrashException` can be propagated directly) and it allows us to return `CompletableFuture<ChunkBuildResult>` matching the return type of `schedule`.

This is what the crash report context looks like:

```
-- Block being rendered --
Details:
	Block: Block{minecraft:gravel}
	Block location: World: (-16,-16,-48), Section: (at 0,0,0 in -1,-1,-3; chunk contains blocks -16,-64,-48 to -1,319,-33), Region: (-1,-1; contains chunks -32,-32 to -1,-1, blocks -512,-64,-512 to -1,319,-1)
	Chunk section: RenderSection at chunk (-1, -1, -3) from (-16, -16, -48) to (0, 0, -32)
	Render context volume: BlockBox{minX=-18, minY=-18, minZ=-50, maxX=1, maxY=1, maxZ=-31}
```

comp500 committed:
Error handling improvements in chunk build threads

Added a queue to store chunk build failures in `ChunkBuilder`
Modified `scheduleDeferred` to add `CompletableFuture` exceptions to this queue
Created `handlePendingFailures` in `RenderSectionManager` to drain the failure queue and throw any found exception
Modified `WorkStealingFutureDrain.findNext` to unwrap `CompletionException`s that contain `CrashException` so that the crash report context is preserved
Wrapped the `ChunkRenderRebuildTask.performBuild` outer loop in a try-catch block to handle exceptions from chunk building and add chunk section / render context information
Improved `RenderSection.toString` to provide more useful information
Changed `ChunkBuilder.processJob` to use the Sodium logger rather than `printStackTrace`

Rebased atop 7aba7c71 on 01/08/23:
Modified `ChunkJobResult.unwrap` to unwrap `CompletionException`s that contain `CrashException` so that the crash report context is preserved
Wrapped the `ChunkRenderMeshingTask.performBuild` outer loop in a try-catch block to handle exceptions from chunk building and add chunk section / render context information
Improved `RenderSection.toString` to provide more useful information
Updated `ChunkJobTyped.execute` to use the Sodium logger (there may be multiple chunk build errors before the main thread gets the results and aborts, so it's useful to log everything)

comp500 committed:
Fix RenderSection bounds in toString changes

jellysquid3 commented:
Can you rebase this against the `/dev` branch? We changed a lot to do with the `ChunkBuilder` but the crash report functionality you added should still work. We also now store a "job output" queue which the main thread polls every frame, and it handles unwrapping any exceptions if they exist.

comp500 commented:
Updated list of changes after rebase:

- ~~Added a queue to store chunk build failures in `ChunkBuilder`~~
- ~~Modified `scheduleDeferred` to add `CompletableFuture` exceptions to this queue~~
- ~~Created `handlePendingFailures` in `RenderSectionManager` to drain the failure queue and throw any found exception~~
- Modified ~~`WorkStealingFutureDrain.findNext`~~ `ChunkJobResult.unwrap` to unwrap `CompletionException`s that contain `CrashException` so that the crash report context is preserved
- Wrapped the `ChunkRenderMeshingTask.performBuild` outer loop in a try-catch block to handle exceptions from chunk building and add chunk section / render context information
- Improved `RenderSection.toString` to provide more useful information
- Updated `ChunkJobTyped.execute` to use the Sodium logger (there may be multiple chunk build errors before the main thread gets the results and aborts, so it's useful to log everything)

jellysquid3 commented on src/main/java/me/jellysquid/mods/sodium/client/render/chunk/compile/executor/ChunkJobResult.java:
```diff
@@ -18,7 +20,10 @@ public static <OUTPUT> ChunkJobResult<OUTPUT> successfully(OUTPUT output) {
 
     public OUTPUT unwrap() {
-        if (this.throwable != null) {
+        if (this.throwable instanceof CrashException) {
```
Should use `if (this.throwable instanceof CrashException crashException)` since it does the instanceof and type-cast in one.

comp500 replied:
Ah, I forgot that was a feature! Been writing Java 8 for too long...

jellysquid3 commented on src/main/java/me/jellysquid/mods/sodium/client/render/chunk/compile/executor/ChunkJobTyped.java:
```diff
@@ -48,6 +49,7 @@ public void execute(ChunkBuildContext context) {
             result = ChunkJobResult.successfully(output);
         } catch (Throwable throwable) {
             result = ChunkJobResult.exceptionally(throwable);
+            SodiumClientMod.logger().error("Chunk build failed", throwable);
```
We've been trying to move away from the global logger and to using class-specific loggers in Sodium. Can you make the logger in `ChunkBuilder` package-private and use it here instead?

jellysquid3 commented on src/main/java/me/jellysquid/mods/sodium/client/render/chunk/compile/tasks/ChunkBuilderMeshingTask.java:
```diff
@@ -139,6 +151,23 @@ public ChunkBuildOutput execute(ChunkBuildContext buildContext, CancellationToke
+        CrashReportSection.addBlockInfo(crashReportSection, slice, pos, state);
+
+        crashReportSection.add("Chunk section", render);
+        if (renderContext != null) {
```
Style nitpick: `this` prefix needed for instance field accesses.

jellysquid3 commented:
Changes look good to me outside of minor nitpicks; will merge when resolved.

comp500 committed:
Fix style issues, use ChunkBuilder logger

Octol1ttle commented on src/main/java/me/jellysquid/mods/sodium/client/render/chunk/compile/tasks/ChunkBuilderMeshingTask.java:
```diff
@@ -101,6 +102,9 @@ public ChunkBuildOutput execute(ChunkBuildContext buildContext, CancellationToke
                             context.update(blockPos, modelOffset, blockState, model, seed);
                             cache.getBlockRenderer()
                                 .renderModel(context, buffers);
+                            if (Random.createLocal().nextInt(100000) == 0) {
+                                throw new RuntimeException("henlo");
+                            }
```
Debug leftovers?

comp500 replied:
Oops...

comp500 committed:
Remove testing code
